lib/vrl/compiler/src/expression/op.rs

```diff
@@ -85,11 +107,16 @@ impl Expression for Op {
 
         let lhs_kind = lhs_def.kind();
         let rhs_kind = rhs_def.kind();
-        let merged_kind = merged_def.kind();
 
         match self.opcode {
-            // null || null
-            Or if merged_kind.is_null() => TypeDef::new().infallible().null(),
+            // ok/err ?? ok
+            Err if rhs_def.is_infallible() => merged_def.infallible(),
+
+            // ... ?? ...
+            Err => merged_def,
+
+            // catch-all to make sure fallible lhs is not ignored
+            _ if lhs_def.is_fallible() => merged_def,
```

Comment:
Can you explain the reasoning here a bit more? 

If the lhs is fallible, then we want to push that up straight away because it will raise an error? Thus all the further cases below which also specify the return types of the operation can be ignored, as errors supercede types?

Just want to make sure I understand this correctly. It may be useful putting a comment here to explain this too because it could well continue to make developers pause at this..

Reply:
Correct. The only operation (thus far) that allows a fallible lhs is `Err`, which is handled above this catch-all.

What this does is basically guard against us forgetting to check for `is_fallible()` in all the following matches, otherwise those aren't working as they should.

Although having said that, I might actually remove this again, now that all the `infallible()` flags have been removed. It's still something we need to be careful about, and even though errors supersede types, it'd still be nice to return the correct types here as well, even if it'll still result in a compiler error because the lhs is fallible.

Reply:
I removed this catch-all 👍